tests/integration/conftest.py

```diff
@@ -1,218 +1,128 @@
+    await model.remove_application(application.name, block_until_done=True)
 
 
-@pytest.fixture(scope="module")
-def app_name(metadata):
-    """Provides app name from the metadata."""
-    yield metadata["name"]
+@pytest_asyncio.fixture(scope="module", name="machine_controller")
+async def machine_controller_fixture() -> typing.AsyncGenerator[Controller, None]:
+    """The lxd controller."""
+    controller = Controller()
+    await controller.connect_controller("localhost")
 
+    yield controller
 
-@pytest_asyncio.fixture(scope="module")
-async def agent_label(ops_test: OpsTest):
-    """Provides label that uniquely identifies the agent under test."""
-    yield ops_test.model_name
+    await controller.disconnect()
 
 
-@pytest_asyncio.fixture(scope="module")
-async def app(
-    ops_test: OpsTest,
-    app_name: str,
-    jenkins_agent_image: str,
-    jenkins_controller_name: str,
-    jenkins_model_name: str,
-    agent_label: str,
-):
-    """Charm used for integration testing.
+@pytest_asyncio.fixture(scope="module", name="machine_model")
+async def machine_model_fixture(
+    machine_controller: Controller,
+) -> typing.AsyncGenerator[Model, None]:
+    """The machine model for jenkins machine charm."""
+    machine_model_name = f"jenkins-server-machine-{secrets.token_hex(2)}"
+    logger.info("Adding model %s on cloud localhost", machine_model_name)
+    model = await machine_controller.add_model(machine_model_name)
 
-    Builds the charm and deploys it and the relations it depends on.
-    """
-    # This helps with type hints, it seems model could be None
-    assert ops_test.model is not None
+    yield model
 
-    # Build and deploy the charm
-    charm = await ops_test.build_charm(".")
-    resources = {"jenkins-agent-image": jenkins_agent_image}
-    application = await ops_test.model.deploy(
-        charm,
-        resources=resources,
-        application_name=app_name,
-        series="jammy",
-        config={"jenkins_agent_labels": agent_label},
-    )
-    await ops_test.model.wait_for_idle()
-
-    juju_model = juju.model.Model()
-    await juju_model.connect("testing")
-    juju_controller = await juju_model.get_controller()
-    await juju_controller.connect()
-    controller_name = juju_controller.controller_name
-
-    assert controller_name is not None
-
-    # Create the relationship
-    jenkins_controller_model_name = f"{jenkins_controller_name}:{jenkins_model_name}"
-    await ops_test.model.create_offer(application_name=app_name, endpoint=f"{app_name}:slave")
-    # [2022-08-26] Cannot use native ops_test functions because they do not support
-    # cross-controller operations
-    await ops_test.juju(
-        "add-relation",
-        "jenkins",
-        f"{controller_name}:{ops_test.model_name}.{app_name}",
-        "--model",
-        jenkins_controller_model_name,
-        check=True,
-    )
-    await ops_test.model.wait_for_idle()
+    await model.disconnect()
```

Comment:
Without a properly terminated connection, the test will spew out a lot of error message at the end of the test, it doesn't affect functionality but aesthetics :D

Reply:
@benhoyt sounds like something we could implement in the library?

Reply:
@jdkandersson Not quite sure what your request is -- that it be automatic rather than explicit?

It looks like it's a documented part of the API, for example [here](https://pythonlibjuju.readthedocs.io/en/latest/api/juju.model.html?highlight=disconnect#juju.model.Model.disconnect) and [use in a try/finally here](https://pythonlibjuju.readthedocs.io/en/latest/#quickstart). @cderici Is the reason this isn't automatic because it's an `async` method, so it can't be done in `__del__` or whatever? Not that that's necessarily better -- explicit it better than implicit. An `async with` might be nice I guess.

Reply:
perhaps a pytest fixture could be provided by the ops testing library that calls disconnect?